### PR TITLE
Prepare for future batching

### PR DESCRIPTION
**Pull request by Thomas-Christie**
## Type of changes

- [ ] Bug fix
- [X] New feature
- [ ] Documentation / docstrings
- [ ] Tests
- [ ] Other

## Checklist

- [X] I've formatted the new code by running `poetry run pre-commit run --all-files --show-diff-on-failure` before committing.
- [X] I've added tests for new code.
- [X] I've added docstrings for the new code.

## Description

In the future certain utility functions will be "batched" utility functions which quantify the utility of querying a *batch* of points rather than a single point. These utility functions may in turn be optimised differently.

Therefore, I propose we have a `SinglePointUtilityFunction` type for non-batched utility functions, which has type annotation `[N  D] -> [N 1]`, and in the future we may add `BatchUtilityFunction` which has type  annotation `[N B D] -> [N 1]`. Distinguishing between these makes it easy for users to see which utility functions support batching.

Also refactored UtilityDrivenDecisionMaker to enable for Thompson sampling to be used in a batched context, and updated unit tests to reflect this. Finally, added a new notebook to serve as an introduction to the `decision_making` module.

Issue Number: N/A

### CODE REVIEW

**Review by thomaspinder — APPROVED** [review]

File: gpjax/decision_making/decision_maker.py [hunk 1]
```diff
@@ -222,28 +240,64 @@ class UtilityDrivenDecisionMaker(AbstractDecisionMaker):
...
+                maximizer = self.utility_maximizer.maximize(
+                    decision_function, self.search_space, key
+                )
+                maximizers.append(maximizer)
+                _, key = jr.split(key)
+
+            maximizers = jnp.concatenate(maximizers)
+            return maximizers
+        else:
+            raise NotImplementedError(
```
Comment (henrymoss) [inline gpjax/decision_making/decision_maker.py:301]: this keeps it tidy for now

File: gpjax/decision_making/decision_maker.py [hunk 2]
```diff
@@ -222,28 +240,64 @@ class UtilityDrivenDecisionMaker(AbstractDecisionMaker):
...
+            Float[Array, "B D"]: Point(s) to be queried next.
         """
-        self.current_utility_function = (
-            self.utility_function_builder.build_utility_function(
-                self.posteriors, self.datasets, key
+        self.current_utility_functions = []
+        maximizers = []
+        if isinstance(self.utility_function_builder, ThompsonSampling) or (
+            (not isinstance(self.utility_function_builder, ThompsonSampling))
+            and (self.batch_size == 1)
```
Comment (henrymoss) [inline gpjax/decision_making/decision_maker.py:279]: does this still not work with batch_size=1? This logic is a little tricky to understand

**Review by henrymoss — APPROVED** [review]
LGTM

File: gpjax/decision_making/decision_maker.py (hunk 2 above)
Comment (Thomas-Christie) [reply gpjax/decision_making/decision_maker.py:279]: It's restricting batch size to be 1 for non-TS acquisition functions if someone happens to implement one for now.

Comment (henrymoss) [reply gpjax/decision_making/decision_maker.py:279]: got it. Maybe you can make the logic clearer so people like me dont get confused?

**Pull request merged**